include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
+
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/any_sender_of.hpp>
+#include <unifex/async_destroy.hpp>
+#include <unifex/async_manual_reset_event.hpp>
+#include <unifex/async_resource_ptr.hpp>
+#include <unifex/defer.hpp>
+#include <unifex/just_done.hpp>
+#include <unifex/just_from.hpp>
+#include <unifex/just_void_or_done.hpp>
+#include <unifex/let_done.hpp>
+#include <unifex/let_value.hpp>
+#include <unifex/manual_lifetime.hpp>
+#include <unifex/nest.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sequence.hpp>
+#include <unifex/spawn_detached.hpp>
+#include <unifex/task.hpp>
+#include <unifex/then.hpp>
+#include <unifex/unstoppable.hpp>
+#include <unifex/v1/async_scope.hpp>
+#include <unifex/when_all.hpp>
+#include <unifex/with_query_value.hpp>
+
+#include <algorithm>
+#include <atomic>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+#include <utility>
+
+namespace unifex {
+namespace _async_resource {
+/*
+* `ResourceFactory` takes an `async_scope_ref` that refers to the *inner* scope
+* `ResourceFactory(async_scope_ref)` returns a `Resource`
+* the `Resource` is stored next to the *inner* scope in a heap-allocated
+  `_container`
+* `async_resource` takes an *outer* scope ref, returns a move-only
+  `async_resource_ptr<Resource>`
+* the returned `async_resource_ptr<Resource>` is a *unique_ptr*-like object
+* creation of the `Resource`, it's asynchronous destruction and `complete()` of
+  the *inner* scope is spawned on the *outer* scope. Lifetime is controlled by
+  `async_resource_ptr<Resource>`.
+* `async_destroy` is a CPO algorithm that by default does nothing. Customizable
+  through `tag_invoke` to do async destruction if `Resource` provides a
+  `destroy() -> Sender`
+* TBD customizations:
+  - for a range of T, async_destroy(range) is just async_destroy on each T,
+  followed by clearing the range
+  - for an optional<T>, async_destroy(opt<T>) is a no-op on empty
+  - async_destroy(*opt) on non-empty, followed by .reset()
+  - for a unique_ptr<T>, async_destroy is async_destroy(*ptr) followed by
+  .reset(), unless it's nullptr, in which case no-op
+*/
+
+class _container_base {
+protected:
+  enum class state_t : char {
+    ALLOCATED,
+    DESTRUCT_SPAWNED,
+    CONSTRUCTED,
+    DESTROYING,
+  };
+
+  void (*destruct_this_resource_)(_container_base*) noexcept;
+  unifex::any_sender_of<> (*schedule_destruct_)(
+      _container_base*, _container_base*) noexcept;
+  unifex::any_sender_of<> (*destroy_this_resource_)(_container_base*) noexcept;
+  void (*deleter_)(_container_base*) noexcept;
```

Comment:
Can these all be `noexcept`?  I think `deleter_` and `destroy_this_resource_` (or `destruct_this_resource_`?) might be since they're basically type-erased destructors.  `schedule_destruct_` is probably a throwing function since `any_sender_of<>` allocates.